contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,553 @@
+not there was an error. The errors may include (but not limited to):
+  * API version not supported
+  * A device plugin already registered this `ResourceName`
+
+Kubelet will then interact with the plugin through the following functions:
+  * ListAndWatch: The device plugin advertises a list of Devices to Kubelet
+    and sends it again if the state of a Device changes
+  * `Allocate`: Called when creating a container with a list of devices
+
+![Process](device-plugin.png)
+
+
+## API Specification
+
+```go
+// Registration is the service advertised by the Kubelet
+// Only when Kubelet answers with a success code to a Register Request
+// may Device Plugins start their service
+service Registration {
+	rpc Register(RegisterRequest) returns (RegisterResponse) {}
+}
+
+// DevicePlugin is the service advertised by Device Plugins
+service DevicePlugin {
+	// ListWatch returns a stream of List of Devices
+	// Whenever a Device state change or a Device disapears, ListWatch
+	// returns the new list
+	rpc ListWatch(Empty) returns (stream ListWatchResponse) {}
+
+	// Allocate is called during container creation so that the Device
+	// Plugin can run device specific operations and instruct Kubelet
+	// of the steps to make the Device available in the container
+	rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+}
+
+message RegisterRequest {
+	// Version of the API the Device Plugin was built against
+	string version = 1;
+	// Name of the unix socket the device plugin is listening on
+	string endpoint = 2;
+	// Name of the Device advertised same as the one in the Device struct
+	string resourceName = 3;
+}
+
+message RegisterResponse {
+	// Minimum version the Kubelet API supports.
+	bool success = 1;
+	// Kubelet fills this field if it encounters any errors
+	// during the registration process
+	string error = 2;
+}
+
+// Allocate is called during container creation so that the Device
+// Plugin can run device specific operations on the Devices sent
+message AllocateRequest {
+	repeated Device devices = 1;
+}
+
+message AllocateResponse {
+	repeated DeviceRuntimeSpec spec = 1;
+}
+
+// ListWatch returns a stream of List of Devices
+// Whenever a Device state change or a Device disapears, ListWatch
+// returns the new list
+message ListWatchResponse {
+	repeated Device devices = 1;
+}
+
+// The list of be added to the CRI spec
+message DeviceRuntimeSpec {
+	// List of environment variable to set in the container.
+	repeated KeyValue envs = 1;
+	// Mounts for the container.
+	repeated Mount mounts = 2;
+}
+
+// E.g:
+// struct Device {
+//    ResourceName: "nvidia-gpu"
+//    ID: "GPU-fef8089b-4820-abfc-e83e-94318197576e",
+//    State: "Healthy",
+//    Properties: {
+//        "Family": "Pascal",
+//        "Memory": "4G",
+//        "ECC"   : "True",
+//    }
+//}
+message Device {
+	string resourceName = 1;
+	string id = 2;
+	string state = 3;
+
+	map<string, string> properties = 4;
```

Comment:
@jiayingz I am working on[ ResourceClass proposal ](https://github.com/kubernetes/community/pull/782) which is about using device properties. I am keeping track with device plugin proposal so that i can rebase ResourceClass proposal accordingly. Therefore, I would like this proposal to keep device properties. May be in simple form with node homogeneity assumption and keep it at register phase.